operators/src/util/raster_stream_to_geotiff.rs

```diff
@@ -0,0 +1,231 @@
+use geoengine_datatypes::{
+    primitives::{SpatialBounded, TimeInterval},
+    raster::{Blit, GeoTransform, Grid2D, GridShape2D, GridSize, Pixel, RasterTile2D},
+    spatial_reference::SpatialReference,
+};
+use std::{
+    convert::TryInto,
+    sync::mpsc::{Receiver, Sender},
+};
+use std::{ffi::CString, sync::mpsc};
+
+use crate::util::Result;
+use crate::{
+    engine::{QueryContext, QueryRectangle, RasterQueryProcessor},
+    error::Error,
+};
+
+pub async fn raster_stream_to_geotiff_bytes<T, C: QueryContext + 'static>(
+    processor: Box<dyn RasterQueryProcessor<RasterType = T>>,
+    query_rect: QueryRectangle,
+    query_ctx: C,
+    width: u32,
+    height: u32,
+    no_data_value: Option<f64>,
+    spatial_reference: SpatialReference,
+) -> Result<Vec<u8>>
+where
+    T: Pixel + GdalType,
+{
+    let file_name = format!("/vsimem/{}.tiff", uuid::Uuid::new_v4());
+    let (tx, rx): (Sender<RasterTile2D<T>>, Receiver<RasterTile2D<T>>) = mpsc::channel();
+
+    let file_name_clone = file_name.clone();
+    let writer = tokio::task::spawn_blocking(move || {
+        gdal_writer(
+            &rx,
+            &file_name_clone,
+            query_rect,
+            width,
+            height,
+            no_data_value,
+            spatial_reference,
+        )
+    });
+
+    let mut tile_stream = processor.raster_query(query_rect, &query_ctx).await?;
+
+    while let Some(tile) = tile_stream.next().await {
+        tx.send(tile?).map_err(|_| Error::ChannelSend)?;
+    }
+
+    drop(tx);
+
+    writer.await??;
+
+    // TODO: use higher level rust-gdal method when it is mapped
+    let bytes = get_vsi_mem_file_bytes_and_free(&file_name);
+
+    Ok(bytes)
+}
+
+fn gdal_writer<T: Pixel + GdalType>(
+    rx: &Receiver<RasterTile2D<T>>,
+    file_name: &str,
+    query_rect: QueryRectangle,
+    width: u32,
+    height: u32,
+    no_data_value: Option<f64>,
+    spatial_reference: SpatialReference,
+) -> Result<()> {
+    let x_pixel_size = query_rect.bbox.size_x() / f64::from(width);
+    let y_pixel_size = query_rect.bbox.size_y() / f64::from(height);
+
+    let output_geo_transform =
+        GeoTransform::new(query_rect.bbox.upper_left(), x_pixel_size, -y_pixel_size);
+    let output_bounds = query_rect.bbox;
+
+    let driver = Driver::get("GTiff")?;
+    // TODO: "COMPRESS, DEFLATE" flags but rust-gdal doesn't support setting this yet(?)
+    let mut dataset =
+        driver.create_with_band_type::<T>(&file_name, width as isize, height as isize, 1)?;
+
+    dataset.set_spatial_ref(&spatial_reference.try_into()?)?;
+    dataset.set_geo_transform(&output_geo_transform.into())?;
+    let mut band = dataset.rasterband(1)?;
+
+    if let Some(no_data) = no_data_value {
+        band.set_no_data_value(no_data)?;
+    }
+
+    while let Ok(tile) = rx.recv() {
+        let tile_info = tile.tile_information();
+
+        let tile_bounds = tile_info.spatial_bounds();
+
+        let mat_tile = if output_bounds.contains_bbox(&tile_bounds) {
```

Comment:
if you return a grid here, you can skip creation of a tile in the else case

Reply:
I need the tile to compute window_size and window further down.

Reply:
but why? the window size is in pixels. and you can calculate the overlap it in the cases or before that.

Reply:
the window_size is in pixels but the **window** is the upper left pixel index where our write starts. This I have to calculate for the case where the tile is not fully contained in the query rectangle.

Reply:
yes but you know that if you do intersection of the tile boxes. thats what the code in blit.rs does

Reply:
8266cd5e3c70066c759d83db427caf5b3a25ac88